### PR TITLE
Modify a process instance by terminating all element instances of a flow scope

### LINKED ISSUE #9691: Modify a process instance by terminating all element instances of a flow scope
## Description

We want to go one step further and allow the termination of all element instances of a flow scope. As a result, the instance of the flow scope should be terminated as well. The flow scope can't be active if it doesn't have any active element instances anymore. 

The termination should be propagated to the flow scope of the terminated element instance recursively to ensure that all flow scopes are terminated if they don't contain any active element instances. The propagation ends on the process instance scope. If the process instance doesn't have any active element instances then the process instance is terminated as well.   

For example:
image

* check the flow scopes after all terminate instructions are applied
* check each flow scope of the terminated element instances
  * terminate the instance of the flow scope (according to #9690) if it has no active element instances (and no pending sequence flows to take)
  * do the same check for the next upper flow scope if the flow scope was terminated
* verify that 
  * the process instance terminates the flow scope if all element instances are terminated
  * the process instance doesn't complete the flow scope if the element instances are terminated (i.e. doesn't take the outgoing sequence flow)
  * the process instance doesn't complete the flow scope if it has pending sequence flows to take (i.e. outgoing sequence flow is taken and activate command is written but not processed)
  * the process instance is terminated if all element instances are terminated

Out of scope:
* termination of all element instances of a process instance that is created by a call activity

Blocked by 
- #9690

### PR DESCRIPTION
**Pull request by saig0**
## Description

* check the element instances of the flow scopes after the termination instructions of the modification command were applied.
* if a flow scope doesn't have any active element instances or pending activations then terminate the flow scope
* propagate the termination to its parent flow scope because it could be the last active element instance in the parent flow scope

## Related issues

closes #9691 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/ModifyProcessInstanceTerminationTest.java [hunk 1]
```diff
@@ -488,6 +491,75 @@ public void shouldNotCompleteFlowScopeIfElementsAreTerminated() {
...
+        RecordToWrite.command()
+            .processInstance(ProcessInstanceIntent.COMPLETE_ELEMENT, elementRecordOfB.getValue())
+            .key(elementRecordOfB.getKey()),
+        RecordToWrite.command().modification(modificationCommand).key(processInstanceKey));
+
+    // then
+    Assertions.assertThat(
+            RecordingExporter.processInstanceRecords()
+                .withProcessInstanceKey(processInstanceKey)
+                .limit("C", ProcessInstanceIntent.ELEMENT_ACTIVATED))
```
Comment (korthout) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/ModifyProcessInstanceTerminationTest.java:541]: //cc @remcowesterhoud this new limit method is being used 🚀

File: engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceModificationProcessor.java [hunk 2]
```diff
@@ -145,7 +151,7 @@ private void executeGlobalVariableInstructions(
                     variableDocument));
   }
 
-  private void terminateElement(
+  private long terminateElement(
```
Comment (korthout) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceModificationProcessor.java]: 🔧 The way I read this, the reason why this method returns the key of the flow scope is that it's useful. But semantically, it makes more sense for this method to return the element instance that was terminated. You could then still retrieve the flow scope key from it between lines 115 and 117, as well as on line 184.

**Review by korthout — APPROVED** [review]
LGTM 👍 Thanks @saig0 this was easy to review

👍 I really like the different tests

🔧 Please have a look at my suggestions before merging.

EDIT: and sorry for the late review. I saw it yesterday, and then it slipped my mind 🙇

File: engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/ModifyProcessInstanceTerminationTest.java [hunk 3]
```diff
@@ -378,6 +379,48 @@ public void shouldTerminateAllElementsOfNestedFlowScope() {
...
+        .processInstance()
+        .withInstanceKey(processInstanceKey)
+        .modification()
+        .terminateElement(elementInstanceKeyOfA)
+        .modify();
+
+    // then
+    assertThatElementIsTerminated(processInstanceKey, "A");
+    assertThatElementIsTerminated(processInstanceKey, "subprocess");
+    assertThatMessageEventSubscriptionIsDeleted(processInstanceKey, elementInstanceKeyOfSubprocess);
```
[the remark this answers is not in the document]
Comment (saig0) [reply engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/ModifyProcessInstanceTerminationTest.java:426]: I would keep it simple for now. It is using the same behavior code. Testing for "all" events would make it more complicated.

**Comment by saig0** [comment]
bors merge

**Pull request merged**